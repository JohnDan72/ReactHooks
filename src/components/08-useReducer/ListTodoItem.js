import React from 'react';
import PropTypes from 'prop-types';
import { Button, Col, FlexboxGrid, List, Toggle } from "rsuite";
import { Icon } from '@rsuite/icons';
import CheckIcon from '@rsuite/icons/Check';
import { BsFillDiamondFill } from 'react-icons/bs';



const propTypes = {
    todo: PropTypes.object.isRequired,
    handleTogle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired,
};

const defaultProps = {};


const ListTodoItem = ({ todo, handleTogle, handleDelete }) => {
    return (
        <List.Item key={todo.id} className="disabled">
            <FlexboxGrid justify="space-between" align="middle">
                <FlexboxGrid.Item as={Col} className={`${todo.done && 'disabled'}`} >
                    <Icon as={BsFillDiamondFill} spin style={{ marginRight: '10px' }} />
                    {todo.desc}
                </FlexboxGrid.Item>

                <FlexboxGrid.Item as={Col}>
                    <Toggle size="md" className="mr-3"
                        checkedChildren={<CheckIcon />}
                        checked={todo.done}
                        onChange={() => handleTogle(todo.id)} />
                    <Button color="red"
                        appearance="ghost"
                        onClick={() => handleDelete(todo.id)}>delete</Button>
                </FlexboxGrid.Item>
            </FlexboxGrid>
        </List.Item>
    );
}

ListTodoItem.propTypes = propTypes;
ListTodoItem.defaultProps = defaultProps;

export default ListTodoItem;