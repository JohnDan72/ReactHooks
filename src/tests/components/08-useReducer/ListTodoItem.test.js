import { shallow } from "enzyme"
import ListTodoItem from "../../../components/08-useReducer/ListTodoItem";


describe('Pruebas ListTodoItem', () => {
    const todo = {
        id: 50 , 
        desc: 'Aprender React',
        done: false
    };
    const [ handleTogle , handleDelete ] = [ jest.fn() , jest.fn()];
    let wrapper = shallow(<ListTodoItem     todo={todo} 
                                            handleDelete={handleDelete} 
                                            handleTogle={handleTogle} />);

    beforeEach(() => {
        wrapper = shallow(<ListTodoItem     todo={todo} 
                                            handleDelete={handleDelete} 
                                            handleTogle={handleTogle} />);
    });

    test('debe renderizar el componente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    

    test('debe llamar la función borrar', () => {
        wrapper.find('Button').simulate('click');
        // wrapper.find('Button').simulate('click',3);
        // wrapper.find('Button').simulate('click',4);
        expect(handleDelete).toHaveBeenCalledTimes(1);
        expect(handleDelete).toHaveBeenCalledWith(todo.id);
    });

    test('debe llamar la función toggle', () => {
        wrapper.find('Toggle').simulate('change', true);

        expect(handleTogle).toHaveBeenCalledTimes(1);
        expect(handleTogle).toHaveBeenCalledWith(todo.id);

    });

    test('debe mostrar correctamente el texto del elemento agregado', () => {
        const gridItems = wrapper.find('FlexboxGridItem');

        // console.log(gridItems.length);
        // console.log(gridItems.at(0).text()); 

        expect(gridItems.at(0).text().includes(todo.desc)).toBe(true);
    });

    test('debe tener la clase disabled si el todo esta en true', () => {
        const todo = { id: 55 , desc: 'Aprender Angular' , done: true };
        const wrapper = shallow(<ListTodoItem   todo={todo} 
                                                handleDelete={handleDelete} 
                                                handleTogle={handleTogle}/>);
        const gridItem = wrapper.find('FlexboxGridItem').at(0)
        // console.log(wrapper.find('FlexboxGridItem').at(0).prop('className'));

        expect(gridItem.prop('className').includes('disabled')).toBe(true);
    })
    
    
})
