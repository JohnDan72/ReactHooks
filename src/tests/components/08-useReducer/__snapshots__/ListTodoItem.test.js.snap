// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas ListTodoItem debe renderizar el componente 1`] = `
<ListItem
  className="disabled"
  key="50"
>
  <FlexboxGrid
    align="middle"
    justify="space-between"
  >
    <FlexboxGridItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "className": [Function],
            "classPrefix": [Function],
            "lg": [Function],
            "lgHidden": [Function],
            "lgOffset": [Function],
            "lgPull": [Function],
            "lgPush": [Function],
            "md": [Function],
            "mdHidden": [Function],
            "mdOffset": [Function],
            "mdPull": [Function],
            "mdPush": [Function],
            "sm": [Function],
            "smHidden": [Function],
            "smOffset": [Function],
            "smPull": [Function],
            "smPush": [Function],
            "xs": [Function],
            "xsHidden": [Function],
            "xsOffset": [Function],
            "xsPull": [Function],
            "xsPush": [Function],
          },
          "render": [Function],
        }
      }
      className="false"
    >
      <Icon
        as={[Function]}
        fill="currentColor"
        height="1em"
        spin={true}
        style={
          Object {
            "marginRight": "10px",
          }
        }
        width="1em"
      />
      Aprender React
    </FlexboxGridItem>
    <FlexboxGridItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "className": [Function],
            "classPrefix": [Function],
            "lg": [Function],
            "lgHidden": [Function],
            "lgOffset": [Function],
            "lgPull": [Function],
            "lgPush": [Function],
            "md": [Function],
            "mdHidden": [Function],
            "mdOffset": [Function],
            "mdPull": [Function],
            "mdPush": [Function],
            "sm": [Function],
            "smHidden": [Function],
            "smOffset": [Function],
            "smPull": [Function],
            "smPush": [Function],
            "xs": [Function],
            "xsHidden": [Function],
            "xsOffset": [Function],
            "xsPull": [Function],
            "xsPush": [Function],
          },
          "render": [Function],
        }
      }
    >
      <Toggle
        checked={false}
        checkedChildren={<Check />}
        className="mr-3"
        onChange={[Function]}
        size="md"
      />
      <Button
        appearance="ghost"
        color="red"
        onClick={[Function]}
      >
        delete
      </Button>
    </FlexboxGridItem>
  </FlexboxGrid>
</ListItem>
`;
